### ISSUE TITLE
Submission for SC consideration: PEP 654 -- Exception Groups and except*

### ISSUE THREAD

**Issue by gvanrossum** [python/steering-council#60]
Please consider PEP 654 for inclusion in Python 3.10.

The PEP has been discussed at length on python-dev and the discussion has come to a natural end (perhaps a steady state :-). Irit has a fully working implementation (though it needs some cleanup) and it would be ready to land in time before 3.10b1 (May 3rd).

/CC: @1st1, @iritkatriel

**Comment by pablogsal** [comment]
Thanks for the submission, @gvanrossum @1st1 and @iritkatriel. I have added it to our agenda for the next meeting(s).

**Comment by iritkatriel** [comment]
We have read Nathaniel's alternative proposal, and we believe that the two approaches are now clear and that they are unreconcilable.  We would like your guidance on how to proceed.

The following are what we see as the main differences:

1. Nathaniel proposes to change the semantics of try-except such that multiple except clauses can execute. try-except is a decades-old feature which has similar semantics in other languages, and we don't know how to evaluate the risks of (a) backwards compatibility breakage; (b) language ergonomics and predictability when breaking away from the semantics in other languages. We added except* because we assumed that such changes to except should not be considered. We would like a clear indication from the SC whether this aspect of the proposal should be discussed further.

2. A primary design goal of Nathaniel's proposal is to make it easy to iterate over an exception group. In an early draft of PEP-654, ExceptionGroup was iterable. We chose to remove that feature in order to de-emphasize iteration as the way to handle exception groups. (We believe that a correct usage pattern will need to do "if there were CancellationErrors in my async task, do X" rather than "for each CancellationError in my async task, do X").   Our point here is that the discussion about iteration is not about how to provide this capability -- PEP 654 ExceptionGroup can provide an iteration API. The question is whether we should, and if it will turn out that we are wrong and this is useful, then adding it to a PEP-654 ExceptionGroup is a matter of implementing an iterator along the lines of the recipe we provide in the PEP. 

3. While the choice of data structure that the interpreter uses internally to represent an exception group is of secondary importance relative to questions of semantics, we wish to point out that Nathaniel did not discuss how the `__context__` and `__cause__` links of exception groups and the exceptions nested in them will be handled. They cannot be flattened like the `__traceback__`s, and this will add major complications or limitations to his design.  The whole point of ExceptionGroups is to make it possible to handle multiple exceptions without loss of error information. The integrity of the exceptions information, including the cause and context links, must be preserved for this to be a robust language feature.  We do not believe that the exception group data structure, which is a tree that has meaningful context/cause information on its internal nodes, can be flattened to a list without loss of information.

Irit, Yury and Guido
